### PR TITLE
backend/enhancement/Customer-Rating-Driver

### PR DESCRIPTION
**Pull request by vsaimanohar**
## Type of Change
backend/enhancement/Customer-Rating-Driver

- [ ] Bugfix
- [ ] New feature
- [X] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

- backend/enhancement/Customer-Rating-Driver

### Additional Changes

- [X] This PR modifies the database schema (database migration added)
- [X] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## How did you test it?
image

## Checklist

- [X] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [X] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Rating.hs [hunk 1]
```diff
@@ -56,24 +57,30 @@ handler merchantId req ride = do
...
+  let newTotalRatingScore = totalRatingScore + ratingValue
+  when (totalRatings == 0) $
     logTagInfo "PersonAPI" "No rating found to calculate"
-  when (ratingCount >= minimumDriverRatesCount) $ do
-    let newAverage = ratingsSum / fromIntegral ratingCount
-    logTagInfo "PersonAPI" $ "New average rating for person " +|| personId ||+ " , rating is " +|| newAverage ||+ ""
-    void $ QP.updateAverageRating personId newAverage
+  if totalRatings >= minimumDriverRatesCount
+    then do
+      let isValidRating = True
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Rating.hs:78]: No need of creating these variables. Can pass the boolean directly

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 2]
```diff
@@ -1180,6 +1183,7 @@ respondQuote (driverId, _) req = do
     searchReq <- QSR.findById searchTry.requestId >>= fromMaybeM (SearchRequestNotFound searchTry.requestId.getId)
     let mbOfferedFare = req.offeredFare
     organization <- CQM.findById searchReq.providerId >>= fromMaybeM (MerchantDoesNotExist searchReq.providerId.getId)
+    logDebug "skldnhasoidbhasidbasdnaslkdmaspdjaposdhasodna"
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:1186]: Need to remove all such logs

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Instances/Person.hs [hunk 3]
```diff
@@ -99,3 +106,8 @@ instance ToTType' BeamP.Person Person where
         BeamP.faceImageId = getId <$> faceImageId,
         BeamP.alternateMobileNumberEncrypted = alternateMobileNumber <&> unEncrypted . (.encrypted)
       }
+
+calculateRating :: Int -> Int -> Maybe Centesimal
+calculateRating totalRatings totalRatingScore
+  | totalRatings > 0 && totalRatingScore > 0 = Just $ fromIntegral totalRatingScore / fromIntegral totalRatings
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Instances/Person.hs:112]: `totalRatings > 0` is enough

File: Backend/dev/migrations/dynamic-offer-driver-app/0265-add-total_ratings-total_rating_socore-is_valid_rating-to-person.sql [hunk 4]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE atlas_driver_offer_bpp.person ADD COLUMN total_ratings int NOT NULL DEFAULT 0;
+ALTER TABLE atlas_driver_offer_bpp.person ADD COLUMN total_rating_score int NOT NULL DEFAULT 0;
+ALTER TABLE atlas_driver_offer_bpp.person ADD COLUMN is_valid_rating boolean NOT NULL DEFAULT FALSE;
```
Comment (0utkarsh) [inline Backend/dev/migrations/dynamic-offer-driver-app/0265-add-total_ratings-total_rating_socore-is_valid_rating-to-person.sql:3]: Have to backfill all these

**Review by 0utkarsh — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Rating.hs (hunk 1 above)
Comment (vsaimanohar) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Rating.hs:78]: hey @0utkarsh , @ritikaJus has asked for more readability of code so lets keep them?

Comment (0utkarsh) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Rating.hs:78]: I don't think it affects the readability in any way. If you still want it that way, you can add comments instead of extra variables

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Rating.hs [hunk 5]
```diff
@@ -70,24 +71,30 @@ handler merchantId req ride = do
...
-  when (ratingCount == 0) $
+  let newRatingsCount = totalRatings + 1
+  let newTotalRatingScore = totalRatingScore + ratingValue
+  when (totalRatings == 0) $
     logTagInfo "PersonAPI" "No rating found to calculate"
-  when (ratingCount >= minimumDriverRatesCount) $ do
-    let newAverage = ratingsSum / fromIntegral ratingCount
-    logTagInfo "PersonAPI" $ "New average rating for person " +|| personId ||+ " , rating is " +|| newAverage ||+ ""
-    void $ QP.updateAverageRating personId newAverage
+  if totalRatings >= minimumDriverRatesCount
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Rating.hs:90]: @vsaimanohar It can be simplified by returning isValidRating from if statement, as rest of the part is same.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Internal/DriverReferee.hs [hunk 6]
```diff
@@ -43,15 +44,17 @@ linkReferee ::
   ( MonadFlow m,
     EsqDBFlow m r,
     CacheFlow m r,
-    EncFlow m r
+    EncFlow m r,
+    HasFlowEnv m r '["appBackendBapInternal" ::: AppBackendBapInternal]
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Internal/DriverReferee.hs:48]: Why is this change required ?

**Review by RitikaHanish — DISMISSED** [review]

File: Backend/app/rider-platform/rider-app-drainer/package.yaml [hunk 7]
```diff
@@ -146,6 +146,7 @@ executables:
     ghc-options:
       - -threaded
       - -rtsopts
+      - -fwhole-archive-hs-libs
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app-drainer/package.yaml:149]: these flags should be there only for local

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/rider-platform/rider-app-drainer/package.yaml (hunk 7 above)
Comment (vsaimanohar) [reply Backend/app/rider-platform/rider-app-drainer/package.yaml:149]: yes, removing redudant .cabal and .yaml changes

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Rating.hs [hunk 8]
```diff
@@ -70,24 +86,27 @@ handler merchantId req ride = do
...
+  let totalRatingScore = fromMaybe 0 mbtotalRatingScore
+  let newRatingsCount = totalRatings + 1
+  let newTotalRatingScore = totalRatingScore + ratingValue
+  when (totalRatings == 0) $
     logTagInfo "PersonAPI" "No rating found to calculate"
-  when (ratingCount >= minimumDriverRatesCount) $ do
-    let newAverage = ratingsSum / fromIntegral ratingCount
-    logTagInfo "PersonAPI" $ "New average rating for person " +|| personId ||+ " , rating is " +|| newAverage ||+ ""
-    void $ QP.updateAverageRating personId newAverage
+  let isValidRating = totalRatings >= minimumDriverRatesCount
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Rating.hs:107]: here it will be newRatingCount instead of totalRatings.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs [hunk 9]
```diff
@@ -191,6 +191,10 @@ createDriverStat currency distanceUnit driverId = do
             totalCoinsConvertedCash = 0.0,
             currency,
             distanceUnit,
+            rating = Nothing,
+            totalRatings = Just 0,
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs:195]: Why did we populate here with 0, it can be nothing

Comment (vsaimanohar) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Lib/DriverScore.hs:195]: those are inital driverstats, so it should be 0 and i am using this logic for in-code backfilling, where only older driver entries which are Nothing, will be refilled rest will be same

**Review by RitikaHanish — APPROVED** [review]

**Review by hkmangla — APPROVED** [review]

**Review by RitikaHanish — APPROVED** [review]
